contrib/vexecutor/nodeVMotion.c

```diff
@@ -0,0 +1,410 @@
+    Assert(node->ps.state->interconnect_context);
+
+    while (!done)
+    {
+        /* grab TupleTableSlot from our child. */
+        outerNode = outerPlanState(node);
+        outerTupleSlot = ExecProcNode(outerNode);
+
+        /* Running in diagnostic mode, we just drop all tuples. */
+        if (Gp_interconnect_type == INTERCONNECT_TYPE_NIL)
+        {
+            if (!TupIsNull(outerTupleSlot))
+                continue;
+
+            return NULL;
+        }
+
+        if (done || TupIsNull(outerTupleSlot))
+        {
+            doSendEndOfStream(motion, node);
+            done = true;
+        }
+        else
+        {
+            doSendTupleBatch(motion, node, outerTupleSlot);
+
+            Gpmon_M_Incr_Rows_Out(GpmonPktFromMotionState(node));
+            setMotionStatsForGpmon(node);
+            CheckSendPlanStateGpmonPkt(&node->ps);
+
+            if (node->stopRequested)
+            {
+                elog(gp_workfile_caching_loglevel, "Motion initiating Squelch walker");
+                /* propagate stop notification to our children */
+                ExecSquelchNode(outerNode);
+                done = true;
+            }
+        }
+#ifdef MEASURE_MOTION_TIME
+        gettimeofday(&time1, NULL);
+
+		node->motionTime.tv_sec += time1.tv_sec - time2.tv_sec;
+		node->motionTime.tv_usec += time1.tv_usec - time2.tv_usec;
+
+		while (node->motionTime.tv_usec < 0)
+		{
+			node->motionTime.tv_usec += 1000000;
+			node->motionTime.tv_sec--;
+		}
+
+		while (node->motionTime.tv_usec >= 1000000)
+		{
+			node->motionTime.tv_usec -= 1000000;
+			node->motionTime.tv_sec++;
+		}
+#endif
+    }
+
+    Assert(node->stopRequested || node->numTuplesFromChild == node->numTuplesToAMS);
+
+    /* nothing else to send out, so we return NULL up the tree. */
+    return NULL;
+}
+
+void
+doSendTupleBatch(Motion * motion, MotionState * node, TupleTableSlot *outerTupleSlot)
+{
+    int16		    targetRoute;
+    HeapTuple       tuple;
+    SendReturnCode  sendRC;
+    ExprContext    *econtext = node->ps.ps_ExprContext;
+
+    /* We got a tuple from the child-plan. */
+    node->numTuplesFromChild++;
+
+    if (motion->motionType == MOTIONTYPE_FIXED)
+    {
+        if (motion->numOutputSegs == 0) /* Broadcast */
+        {
+            targetRoute = BROADCAST_SEGIDX;
+        }
+        else /* Fixed Motion. */
+        {
+            Assert(motion->numOutputSegs == 1);
+            /*
+             * Actually, since we can only send to a single output segment
+             * here, we are guaranteed that we only have a single
+             * targetRoute setup that we could possibly send to.  So we
+             * can cheat and just fix the targetRoute to 0 (the 1st
+             * route).
+             */
+            targetRoute = 0;
+        }
+    }
+    else if (motion->motionType == MOTIONTYPE_HASH) /* Redistribute */
+    {
+        //TODO:: Implement later
```

Comment:
It's better to error out here than just ignore this condition.

Reply:
VE conditional check in init hook close "vmotion" when it is redistribution type. I literally leave a room for next patch